lib/relooper/src/graph/cfg/mod.rs

```diff
@@ -0,0 +1,175 @@
+            Terminal => Terminal,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Cfg<TLabel> {
+    pub(crate) entry: TLabel,
+    out_edges: HashMap<TLabel, CfgEdge<TLabel>>,
+}
+
+impl<T> Cfg<T> {
+    pub fn edges(&self) -> &HashMap<T, CfgEdge<T>> {
+        &self.out_edges
+    }
+}
+
+impl<T: Eq + Hash + Clone> Cfg<T> {
+    pub fn new(entry: T) -> Cfg<T> {
+        Self {
+            entry,
+            out_edges: Default::default(),
+        }
+    }
+
+    pub fn map_label<'a, M, U: Eq + Hash>(&'a self, mapping: M) -> Cfg<U>
+    where
+        M: Fn(&'a T) -> U,
+    {
+        let out_edges: HashMap<U, CfgEdge<U>> = self
+            .out_edges
+            .iter()
+            .map(|(from, e)| (mapping(from), e.map(&mapping)))
+            .collect();
+
+        Cfg {
+            entry: mapping(&self.entry),
+            out_edges,
+        }
+    }
+
+    pub fn to_borrowed(&self) -> Cfg<&T> {
+        self.map_label(|l| l)
+    }
+
+    pub fn nodes(&self) -> HashSet<&T> {
+        self.out_edges.keys().collect()
+    }
+
+    pub fn children(&self, label: &T) -> HashSet<&T> {
+        self.out_edges
+            .get(label)
+            .into_iter()
+            .flat_map(|edge| edge.to_vec())
+            .collect()
+    }
+
+    pub fn add_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let nodes = self.nodes();
+        // todo unncecessary collect
+        for n in edge
+            .to_vec()
+            .into_iter()
+            .filter(|n| !nodes.contains(n))
+            .collect::<Vec<_>>()
+        {
+            // todo clone
+            self.add_node(n.clone());
+        }
+
+        let prev_edge = self.out_edges.insert(from, edge);
+        match prev_edge {
+            None | Some(Terminal) => {}
+            _ => panic!("adding edge over already present one"),
+        }
+    }
+
+    pub fn add_node(&mut self, n: T) {
+        self.out_edges.insert(n, Terminal);
+    }
+
+    pub fn remove_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let removed_edge = self.out_edges.remove(&from);
+        assert!(removed_edge == Some(edge));
+    }
+
+    pub fn add_edge_or_promote(&mut self, from: T, to: T) {
+        match self.out_edges.remove(&from) {
+            None | Some(Terminal) => self.out_edges.insert(from, Uncond(to)),
+            Some(Uncond(uncond)) => self.out_edges.insert(from, Cond(to, uncond)),
+            _ => panic!("edge (should be absent) or (shouldn't be `Cond`)"),
+        };
+    }
+
+    pub fn edge(&self, label: &T) -> &CfgEdge<T> {
+        self.out_edges
+            .get(label)
+            .expect("any node should have outgoing edges")
+    }
+}
+
+impl<TLabel: Eq + Hash + Copy> Cfg<TLabel> {
+    pub fn from_edges(entry: TLabel, edges: &HashMap<TLabel, CfgEdge<TLabel>>) -> Self {
+        let mut cfg = Cfg::new(entry);
+        for (&from, &edge) in edges.iter() {
+            cfg.add_edge(from, edge);
+        }
+
+        cfg
+    }
+
+    #[deprecated]
```

Comment:
I don't think we should introduce new code that is already deprecated. If we are not using it any more then simply remove it. It's not like we have other people who might be using this library because we just wrote it.

Reply:
I think it's fixed